pkg/kubelet/kubelet_test.go

```diff
@@ -2200,20 +2198,77 @@ func TestMakeEnvironmentVariables(t *testing.T) {
 					},
 				},
 			},
-			"nothing",
-			true,
-			util.NewStringSet(
+			masterServiceNs: "nothing",
+			nilLister:       true,
+			expectedEnvs: util.NewStringSet(
 				"POD_NAME=dapi-test-pod-name",
 				"POD_NAMESPACE=downward-api",
 			),
-			2,
+		},
+		{
+			name: "env expansion",
+			ns:   "test1",
+			container: &api.Container{
+				Env: []api.EnvVar{
+					{
+						Name:  "TEST_LITERAL",
+						Value: "test-test-test",
+					},
+					{
+						Name: "POD_NAME",
+						ValueFrom: &api.EnvVarSource{
+							FieldRef: &api.ObjectFieldSelector{
+								APIVersion: "v1beta3",
+								FieldPath:  "metadata.name",
+							},
+						},
+					},
+					{
+						Name:  "POD_NAME_TEST",
+						Value: "test-$POD_NAME",
+					},
+					{
+						Name:  "POD_NAME_TEST2",
+						Value: "test2-${POD_NAME}",
+					},
+					{
+						Name:  "POD_NAME_TEST3",
+						Value: "test3-${POD_NAME", // Note: brace missing is intentional
+					},
+					{
+						Name:  "LITERAL_TEST",
+						Value: "literal-${TEST_LITERAL}",
+					},
+					{
+						Name:  "SERVICE_VAR_TEST",
+						Value: "${TEST_SERVICE_HOST}:${TEST_SERVICE_PORT}",
+					},
+				},
+			},
+			masterServiceNs: "nothing",
+			nilLister:       false,
+			expectedEnvs: util.NewStringSet(
+				"TEST_LITERAL=test-test-test",
+				"POD_NAME=dapi-test-pod-name",
+				"POD_NAME_TEST=test-dapi-test-pod-name",
+				"POD_NAME_TEST2=test2-dapi-test-pod-name",
+				"POD_NAME_TEST3=test3-POD_NAME",
```

Comment:
@smarterclayton os.Env transforms `${` into the empty string, even if it's unmatched.